File: backend/app/rules/cyberpunk_red/schema.py
from pydantic import BaseModel, ConfigDict, Field

CYBERPUNK_STATS = (
    "int",
    "ref",
    "tech",
    "cool",
    "attr",
    "luck",
    "move",
    "body",
    "emp",
)


class StatBlock(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    intelligence: int = Field(default=4, ge=1, le=10, validation_alias="int", serialization_alias="int")
    ref: int = Field(default=4, ge=1, le=10)
    tech: int = Field(default=4, ge=1, le=10)
    cool: int = Field(default=4, ge=1, le=10)
    attr: int = Field(default=4, ge=1, le=10)
    luck: int = Field(default=4, ge=1, le=10)
    move: int = Field(default=4, ge=1, le=10)
    body: int = Field(default=4, ge=1, le=10)
    emp: int = Field(default=4, ge=1, le=10)

    def score(self, stat: str) -> int:
        key = stat.strip().lower()
        if key not in CYBERPUNK_STATS:
            raise ValueError(f"Unknown stat: {stat}")
        if key == "int":
            return self.intelligence
        return getattr(self, key)


class SkillEntry(BaseModel):
    name: str
    stat: str = "ref"
    rank: int = Field(default=0, ge=0, le=10)


class HpBlock(BaseModel):
    max: int = Field(default=40, ge=1)
    current: int = Field(default=40, ge=0)


class ArmorBlock(BaseModel):
    head: int = Field(default=0, ge=0, le=25)
    body: int = Field(default=0, ge=0, le=25)


class WeaponEntry(BaseModel):
    name: str
    stat: str = "ref"
    skill: str = "handgun"
    damage_dice: str = "2d6"
    damage_type: str = "ballistic"


class CyberpunkRedSheet(BaseModel):
    stats: StatBlock = Field(default_factory=StatBlock)
    skills: list[SkillEntry] = Field(default_factory=list)
    hp: HpBlock = Field(default_factory=HpBlock)
    armor: ArmorBlock = Field(default_factory=ArmorBlock)
    weapons: list[WeaponEntry] = Field(default_factory=list)


SUPPORTED_ROLL_TYPES = [
    "skill_check",
    "stat_check",
    "attack_roll",
    "damage",
    "initiative",
]

File: frontend/src/features/character-sheet/systems/cyberpunk_red/schema.ts
import { z } from "zod";

export const CYBERPUNK_STATS = [
  "int",
  "ref",
  "tech",
  "cool",
  "attr",
  "luck",
  "move",
  "body",
  "emp",
] as const;

export type CyberpunkStat = (typeof CYBERPUNK_STATS)[number];

export const CYBERPUNK_STAT_LABELS: Record<CyberpunkStat, string> = {
  int: "INT",
  ref: "REF",
  tech: "TECH",
  cool: "COOL",
  attr: "ATTR",
  luck: "LUCK",
  move: "MOVE",
  body: "BODY",
  emp: "EMP",
};

const statScore = z.number().int().min(1).max(10);

const skillSchema = z.object({
  name: z.string().min(1, "Nombre requerido"),
  stat: z.enum(CYBERPUNK_STATS),
  rank: z.number().int().min(0).max(10),
});

export const cyberpunkRedSheetSchema = z.object({
  stats: z.object({
    int: statScore,
    ref: statScore,
    tech: statScore,
    cool: statScore,
    attr: statScore,
    luck: statScore,
    move: statScore,
    body: statScore,
    emp: statScore,
  }),
  skills: z.array(skillSchema),
  hp: z.object({
    max: z.number().int().min(1),
    current: z.number().int().min(0),
  }),
  armor: z.object({
    head: z.number().int().min(0).max(25),
    body: z.number().int().min(0).max(25),
  }),
});

export type CyberpunkRedSheet = z.infer<typeof cyberpunkRedSheetSchema>;

export function defaultCyberpunkRedSheet(): CyberpunkRedSheet {
  return {
    stats: {
      int: 4,
      ref: 4,
      tech: 4,
      cool: 4,
      attr: 4,
      luck: 4,
      move: 4,
      body: 4,
      emp: 4,
    },
    skills: [],
    hp: { max: 40, current: 40 },
    armor: { head: 0, body: 0 },
  };
}

export function parseCyberpunkRedSheet(raw: unknown): CyberpunkRedSheet {
  const result = cyberpunkRedSheetSchema.safeParse(raw);
  if (result.success) return result.data;
  return defaultCyberpunkRedSheet();
}
